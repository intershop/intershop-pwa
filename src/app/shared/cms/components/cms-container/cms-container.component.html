<div [ngClass]="containerClasses" class="content-container">
  <ish-content-slot [wrapper]="true" [slot]="'app_sf_base_cm:slot.container.content.pagelet2-Slot'" [pagelet]="pagelet">
    <ish-content-pagelet
      *ngFor="let slotPagelet of contentSlotPagelets"
      [pageletId]="slotPagelet"
    ></ish-content-pagelet>
  </ish-content-slot>
</div>

<!-- TODO: (CMS Component Configuration)
  TrackingCode
  TrackingScript
-->
