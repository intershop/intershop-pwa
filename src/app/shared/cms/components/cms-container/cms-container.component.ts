import { ChangeDetectionStrategy, Component, Input, OnChanges } from '@angular/core';

import { ContentPageletView } from 'ish-core/models/content-view/content-view.model';
import { CMSComponent } from 'ish-shared/cms/models/cms-component/cms-component.model';

@Component({
  selector: 'ish-cms-container',
  templateUrl: './cms-container.component.html',
  changeDetection: ChangeDetectionStrategy.OnPush,
})
export class CMSContainerComponent implements CMSComponent, OnChanges {
  @Input() pagelet: ContentPageletView;

  contentSlotPagelets: string[] = [];
  containerClasses = '';

  ngOnChanges() {
    let contentSlotPagelets = this.pagelet.slot('app_sf_base_cm:slot.container.content.pagelet2-Slot').pageletIDs;
    if (this.pagelet.hasParam('UpperBound')) {
      contentSlotPagelets = contentSlotPagelets.slice(0, this.pagelet.numberParam('UpperBound'));
    }
    this.contentSlotPagelets = contentSlotPagelets;

    this.containerClasses = this.getGridCSS(this.pagelet.stringParam('Grid'));
    this.containerClasses += this.pagelet.stringParam('CSSClass', '');
  }

  getGridCSS(grid: string): string {
    let gridCSS = '';

    if (grid) {
      // transform an incoming string like "ExtraSmall:12,Small:6,Medium:4,Large:0" to a grid object
      const gridObject = { ExtraSmall: 0, Small: 0, Medium: 0, Large: 0 };
      grid.split(',').forEach(element => {
        gridObject[element.split(':')[0] as keyof typeof gridObject] = Number(element.split(':')[1]);
      });

      // the 'hidden-__' classes replacement is not 100% compatible to Bootstrap 3
      if (gridObject.ExtraSmall !== 0) {
        gridCSS += gridObject.ExtraSmall !== -1 ? `col-${gridObject.ExtraSmall} ` : 'd-none d-md-block ';
      }
      if (gridObject.Small !== 0) {
        gridCSS += gridObject.Small !== -1 ? `col-md-${gridObject.Small} ` : 'd-md-none d-lg-block ';
      }
      if (gridObject.Medium !== 0) {
        gridCSS += gridObject.Medium !== -1 ? `col-lg-${gridObject.Medium} ` : 'd-lg-none d-xl-block ';
      }
      if (gridObject.Large !== 0) {
        gridCSS += gridObject.Large !== -1 ? `col-xl-${gridObject.Large} ` : 'd-xl-none ';
      }

      // 'float-left' is added to get the similar styling as with Bootstrap 3
      if (gridCSS !== '') {
        gridCSS += 'float-left ';
      }
    }

    return gridCSS;
  }
}
