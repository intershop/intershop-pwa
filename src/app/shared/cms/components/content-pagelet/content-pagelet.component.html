<ish-content-design-view-wrapper [pageletId]="pageletId">
  <ng-container #cmsOutlet />
</ish-content-design-view-wrapper>
