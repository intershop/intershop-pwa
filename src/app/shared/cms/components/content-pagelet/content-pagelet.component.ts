import {
  ChangeDetectionStrategy,
  ChangeDetectorRef,
  Component,
  ComponentFactoryResolver,
  Injector,
  Input,
  OnChanges,
  OnDestroy,
  OnInit,
  SimpleChange,
  ViewChild,
  ViewContainerRef,
} from '@angular/core';
import { ReplaySubject, Subject } from 'rxjs';
import { switchMap, takeUntil } from 'rxjs/operators';

import { CMSFacade } from 'ish-core/facades/cms.facade';
import { ContentPageletView } from 'ish-core/models/content-view/content-view.model';
import { whenTruthy } from 'ish-core/utils/operators';
import { CMSComponentProvider, CMS_COMPONENT } from 'ish-shared/cms/configurations/injection-keys';
import { CMSComponent } from 'ish-shared/cms/models/cms-component/cms-component.model';
import { SfeAdapterService } from 'ish-shared/cms/sfe-adapter/sfe-adapter.service';
import { SfeMetadataWrapper } from 'ish-shared/cms/sfe-adapter/sfe-metadata-wrapper';
import { SfeMapper } from 'ish-shared/cms/sfe-adapter/sfe.mapper';

/**
 * The Content Pagelet Container Component renders the pagelet for the given 'pageletId'.
 * For the rendering an Angular component is used that has to be provided
 * for the DefinitionQualifiedName of the pagelet.
 *
 * @example
 * <ish-content-pagelet [pageletId]="pagelet"></ish-content-pagelet>
 */
@Component({
  selector: 'ish-content-pagelet',
  templateUrl: './content-pagelet.component.html',
  changeDetection: ChangeDetectionStrategy.OnPush,
})
export class ContentPageletComponent extends SfeMetadataWrapper implements OnChanges, OnInit, OnDestroy {
  /**
   * The Id of the Pagelet that is to be rendered.
   */
  @Input() pageletId: string;

  @ViewChild('cmsOutlet', { read: ViewContainerRef, static: true }) cmsOutlet: ViewContainerRef;

  private pageletId$ = new ReplaySubject<string>(1);
  private destroy$ = new Subject();

  constructor(
    private injector: Injector,
    private componentFactoryResolver: ComponentFactoryResolver,
    private sfeAdapter: SfeAdapterService,
    private cmsFacade: CMSFacade,
    private cdRef: ChangeDetectorRef
  ) {
    super();
  }

  ngOnDestroy() {
    this.destroy$.next();
    this.destroy$.complete();
  }

  ngOnInit() {
    this.pageletId$
      .pipe(
        switchMap(pageletId => this.cmsFacade.pagelet$(pageletId)),
        whenTruthy(),
        takeUntil(this.destroy$)
      )
      .subscribe(pagelet => {
        this.mapComponent(pagelet);
        if (this.sfeAdapter.isInitialized()) {
          this.setSfeMetadata(SfeMapper.mapPageletViewToSfeMetadata(pagelet));
        }
        this.cdRef.markForCheck();
      });
  }

  ngOnChanges() {
    this.pageletId$.next(this.pageletId);
  }

  private mapComponent(pagelet: ContentPageletView) {
    const components = this.injector.get<CMSComponentProvider[]>(CMS_COMPONENT, []);
    const mappedComponent = components.find(c => c.definitionQualifiedName === pagelet.definitionQualifiedName);

    if (mappedComponent) {
      const componentRef = this.createComponent(mappedComponent);
      this.initializeComponent(componentRef.instance, pagelet);
    } else {
      console.warn(`did not find mapping for ${pagelet.id} (${pagelet.definitionQualifiedName})`);
    }
  }

  private createComponent(mappedComponent: CMSComponentProvider) {
    const factory = this.componentFactoryResolver.resolveComponentFactory(mappedComponent.class);
    this.cmsOutlet.clear();
    return this.cmsOutlet.createComponent(factory);
  }

  private initializeComponent(instance: CMSComponent, pagelet: ContentPageletView) {
    instance.pagelet = pagelet;

    // OnChanges has to be manually invoked on dynamically created components
    if (instance.ngOnChanges) {
      instance.ngOnChanges({ pagelet: new SimpleChange(undefined, pagelet, true) });
    }
  }
}
