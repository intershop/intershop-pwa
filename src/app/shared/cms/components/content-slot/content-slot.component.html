<ng-content *ngIf="wrapper; else renderPagelets"></ng-content>

<ng-template #renderPagelets>
  <ish-content-pagelet
    *ngFor="let pagelet of pagelet.slot(slot).pageletIDs"
    [pageletId]="pagelet"
    [callParameters]="callParameters"
  ></ish-content-pagelet>
</ng-template>
