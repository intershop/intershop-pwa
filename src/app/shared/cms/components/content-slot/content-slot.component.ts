import { ChangeDetectionStrategy, Component, Input, OnChanges } from '@angular/core';

import { ContentPageletView } from 'ish-core/models/content-view/content-view.model';
import { SfeAdapterService } from 'ish-shared/cms/sfe-adapter/sfe-adapter.service';
import { SfeMetadataWrapper } from 'ish-shared/cms/sfe-adapter/sfe-metadata-wrapper';
import { SfeMapper } from 'ish-shared/cms/sfe-adapter/sfe.mapper';

/**
 * The Content Slot Container Component renders the assigned sub pagelets
 * of the identified 'slot' of the given 'pagelet'.
 * By default it is just using the {@link ContentPageletContainerComponent}
 * to render each sub pagelet.
 * If more specific HTML or functionality is needed for the rendering
 * the 'wrapper' flag needs to be used to use the provided HTML for rendering.
 *
 * @example
 * <ish-content-slot [slot]="'app_pwa:slot.pagelet2-Slot'" [pagelet]="pagelet"></ish-content-slot>
 *
 * <ish-content-slot [wrapper]="true" [slot]="'app_pwa:slot.pagelet2-Slot'" [pagelet]="pagelet">
 *               <div *ngFor="let slotPagelet of slotPagelets">
 *                 <ish-content-pagelet [pageletId]="slotPagelet"></ish-content-pagelet>
 *               </div>
 * </ish-content-slot>
 */
@Component({
  selector: 'ish-content-slot',
  templateUrl: './content-slot.component.html',
  changeDetection: ChangeDetectionStrategy.OnPush,
})
export class ContentSlotComponent extends SfeMetadataWrapper implements OnChanges {
  /**
   * The DefinitionQualifiedName of the slot that should be rendered.
   */
  @Input() slot: string;
  /**
   * The parent Pagelet that contains the slot.
   */
  @Input() pagelet: ContentPageletView;
  /**
   * An optional flag that controls the rendering of the pagelets with the wrapped HTML content.
   */
  @Input() wrapper?: boolean;

  constructor(private sfeAdapter: SfeAdapterService) {
    super();
  }

  ngOnChanges() {
    if (this.pagelet && this.sfeAdapter.isInitialized()) {
      this.setSfeMetadata(SfeMapper.mapSlotViewToSfeMetadata(this.pagelet, this.slot));
    }
  }
}
