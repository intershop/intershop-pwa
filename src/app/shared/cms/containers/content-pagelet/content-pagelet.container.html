<ng-container #cmsOutlet></ng-container>
