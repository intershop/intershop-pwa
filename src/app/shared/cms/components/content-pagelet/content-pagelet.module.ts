import { Injector, NgModule } from '@angular/core';

import { ModuleLoaderService } from 'ish-core/utils/module-loader/module-loader.service';
import { ContentPageletComponent } from 'ish-shared/cms/components/content-pagelet/content-pagelet.component';

const exportedComponents = [ContentPageletComponent];

@NgModule({
  declarations: [...exportedComponents],
  exports: [...exportedComponents],
})
export class ContentPageletModule {
  constructor(moduleLoader: ModuleLoaderService, injector: Injector) {
    moduleLoader.init(injector);
  }
}
