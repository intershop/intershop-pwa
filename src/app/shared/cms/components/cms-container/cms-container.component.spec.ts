import { ComponentFixture, TestBed, async } from '@angular/core/testing';
import { MockComponent } from 'ng-mocks';

import { createContentPageletView } from 'ish-core/models/content-view/content-view.model';
import { ContentPageletComponent } from 'ish-shared/cms/components/content-pagelet/content-pagelet.component';
import { ContentSlotComponent } from 'ish-shared/cms/components/content-slot/content-slot.component';

import { CMSContainerComponent } from './cms-container.component';

describe('Cms Container Component', () => {
  let component: CMSContainerComponent;
  let fixture: ComponentFixture<CMSContainerComponent>;
  let element: HTMLElement;

  beforeEach(async(() => {
    TestBed.configureTestingModule({
      declarations: [
        CMSContainerComponent,
        MockComponent(ContentPageletComponent),
        MockComponent(ContentSlotComponent),
      ],
    }).compileComponents();
  }));

  beforeEach(() => {
    fixture = TestBed.createComponent(CMSContainerComponent);
    component = fixture.componentInstance;
    const pagelet = {
      definitionQualifiedName: 'fq',
      id: 'id',
      domain: 'domain',
      displayName: 'pagelet',
      configurationParameters: {
        CSSClass: 'foo-class',
        Grid: 'ExtraSmall:12,Small:6,Medium:4,Large:0',
      },
      slots: [
        {
          displayName: 'test',
          definitionQualifiedName: 'app_sf_base_cm:slot.container.content.pagelet2-Slot',
          pageletIDs: ['slide1', 'slide2'],
        },
      ],
    };
    component.pagelet = createContentPageletView(pagelet);
    element = fixture.nativeElement;
  });

  it('should be created', () => {
    expect(component).toBeTruthy();
    expect(element).toBeTruthy();
    expect(() => component.ngOnChanges()).not.toThrow();
    expect(() => fixture.detectChanges()).not.toThrow();
    expect(element).toMatchInlineSnapshot(`
      <div
        class="content-container col-12 col-md-6 col-lg-4 float-left foo-class"
        ng-reflect-ng-class="col-12 col-md-6 col-lg-4 float"
      >
        <ish-content-slot ng-reflect-slot="app_sf_base_cm:slot.container." ng-reflect-wrapper="true"
          ><ish-content-pagelet ng-reflect-pagelet-id="slide1"></ish-content-pagelet
          ><ish-content-pagelet ng-reflect-pagelet-id="slide2"></ish-content-pagelet
        ></ish-content-slot>
      </div>
    `);
  });
});
