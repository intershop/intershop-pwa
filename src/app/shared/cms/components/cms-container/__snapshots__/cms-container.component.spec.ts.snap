// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Cms Container Component should be created 1`] = `
<div
  class="content-container col-12 col-md-6 col-lg-4 float-left foo-class"
  ng-reflect-klass="content-container"
  ng-reflect-ng-class="col-12 col-md-6 col-lg-4 float"
>
  <ish-content-slot
    ng-reflect-slot="app_sf_responsive_cm:slot.cont"
    ng-reflect-pagelet="[object Object]"
    ng-reflect-wrapper="true"
    ><ish-content-pagelet ng-reflect-pagelet="[object Object]"></ish-content-pagelet
    ><ish-content-pagelet ng-reflect-pagelet="[object Object]"></ish-content-pagelet
  ></ish-content-slot>
</div>
`;
